### PR TITLE
[ShapeOpt] omp parallel assembly of symmetric mapping matrix"

### PR DESCRIPTION
**Pull request by armingeiser**
**Description**
Follow up from #8676 omp parallel assembly of symmetric mapping matrix.

**Changelog**
-  omp parallel assembly of symmetric mapping matrix.

### CODE REVIEW

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp [hunk 1]
```diff
@@ -218,6 +221,7 @@ void MapperVertexMorphingSymmetric::ComputeMappingMatrix()
 
     // fixed size vec
     NodeVectorType neighbor_nodes( max_number_of_neighbors );
+    #pragma omp for
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp]: this should use the parallelUtils with ThreadLocalStorage I strongly recommend (https://github.com/KratosMultiphysics/Kratos/wiki/Parallel-Utilities#basic-statically-scheduled-for-loop)

However IIRC there is a bit more to do, e.g. making sure inserting in the MappingMatrix should be atomic etc

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp]: At this stage i know that each thread will only assemble to a certain row, so there will never be two threads assigning to the same row or even element in the matrix. That's why i never explicitly made the assignment atomic. But i guess that is bad practice?

Comment (philbucher) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp]: ah ok

well if you know this then I would not add the atomic, but a comment somewhere

If you can write the algorithm without need for synchronization is better IMO as faster

I took a look at the file, actually it seems to be ok as is!

I.e. if you want you could even merge as is. I recommend using the parallel utils but is is up to you

**Review by philbucher — DISMISSED** [review]

**Review by philbucher — DISMISSED** [review]
:+1:

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp [hunk 2]
```diff
@@ -206,51 +207,58 @@ void MapperVertexMorphingSymmetric::ComputeMappingMatrix()
...
+        std::vector<bool> transform;
+        NodeVectorType total_neighbor_nodes;
+        std::vector<double> total_list_of_weights;
+        std::vector<double> list_of_weights;
+
+        // fixed size vec
+        NodeVectorType neighbor_nodes;
+
+        // constructor
+        tls_vecs(const unsigned int MaxNumberOfNeighbors) {
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp]:
```suggestion
        explicit tls_vecs(const unsigned int MaxNumberOfNeighbors) {
```

**Review by philbucher — APPROVED** [review]

**Pull request merged**